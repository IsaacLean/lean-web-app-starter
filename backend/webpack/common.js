/**
 * This is the webpack configuration that is used for all modes.
 */
const path = require('path')
const setupNodeExternals = require('ljas-webpack/setupNodeExternals')
const { compileTS } = require('ljas-webpack')
const { merge } = require('webpack-merge')

module.exports = merge([
  {
    entry: './src/main.ts',
  },

  compileTS({ include: path.resolve(__dirname, '../src') }),

  setupNodeExternals(),
])
