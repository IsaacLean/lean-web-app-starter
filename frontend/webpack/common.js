/**
 * This is the webpack configuration that is used for all modes.
 */
const buildHTML = require('ljas-webpack/buildHTML')
const path = require('path')
const { compileReact } = require('ljas-webpack')
const { merge } = require('webpack-merge')

module.exports = merge([
  {
    entry: './src/main.tsx',
  },

  buildHTML('Lean JavaScript Application Starter', 'src/index.html'),

  compileReact(path.resolve(__dirname, '../src')),
])
