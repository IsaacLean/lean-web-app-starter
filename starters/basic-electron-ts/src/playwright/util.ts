import {
  ElectronApplication,
  Page,
  _electron as electron,
} from "@playwright/test";

/**
 * Display console API method calls from the Electron application in the Playwright console.
 * @param electronApp Electron application representation
 */
export const enableConsole = (electronApp: ElectronApplication) => {
  electronApp.on("console", (data) => console.log(data));
};

/**
 * Get the operating system that the browser is running in.
 * @param page Electron Page instance
 * @returns A promise that will resolve to the Browser's operating system type, or null if
 *     it could not be determined
 */
export const getBrowserOs = async (page: Page) => {
  const userAgent = await page.evaluate(async () => window.navigator.userAgent);

  if (userAgent.includes("Windows")) {
    return "win";
  } else if (userAgent.includes("Mac")) {
    return "mac";
  } else if (userAgent.includes("X11")) {
    return "unix";
  } else if (userAgent.includes("Linux")) {
    return "linux";
  }

  return null;
};

/**
 * Launch the Electron application with either the development or production build.
 * By default the PLAYWRIGHT_BUILD_TYPE environment variable will determine which build will be used,
 * but you can override this through some of the available parameters.
 * @param arg Parameter object
 * @param arg.buildPath Manually set the build path. Overrides buildType parameter and PLAYWRIGHT_BUILD_TYPE environment variable.
 * @param arg.buildType Determines if the development or the production build is used for testing. Overrides PLAYWRIGHT_BUILD_TYPE environment variable.
 * @param arg.launchOptions Options to pass to Electron.launch method
 * @returns A promise that will resolve to an Electron application representation
 */
export const launchElectron = ({
  buildPath,
  buildType,
  launchOptions,
}: {
  buildPath?: string;
  buildType?: "development" | "production";
  launchOptions?: Parameters<typeof electron.launch>[0];
} = {}) => {
  let arg = ".";
  if (buildPath) {
    arg = buildPath;
  } else if (
    buildType === "production" ||
    process.env.PLAYWRIGHT_BUILD_TYPE === "production"
  ) {
    arg = "./build/production/main/main.js";
  }

  return electron.launch({
    args: [arg],
    ...launchOptions,
  });
};
