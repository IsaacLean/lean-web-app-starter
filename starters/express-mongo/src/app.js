import compression from "compression";
import express from "express";

const app = express();

if (process.env.NODE_ENV === "production") {
  app.use(compression()); // Middleware that compresses most response bodies
}

app.get("/", (req, res) => res.send("Hello World!"));

export default app;
