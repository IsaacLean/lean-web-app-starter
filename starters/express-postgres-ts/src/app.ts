import compression from 'compression'
import express from 'express'
import helmet from 'helmet'

const app = express()

app.use(helmet()) // Middleware that enhances security by setting HTTP response headers

if (process.env.NODE_ENV === 'production') {
    app.use(compression()) // Middleware that compresses most response bodies
}

app.get('/', (req, res) => res.send('Hello World!'))

export default app
