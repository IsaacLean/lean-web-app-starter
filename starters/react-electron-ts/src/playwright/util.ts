import {
    ElectronApplication,
    Page,
    _electron as electron,
} from '@playwright/test'

/**
 * Display console API method calls from the Electron application in the Playwright console.
 * @param electronApp Electron application representation
 */
export const enableConsole = (electronApp: ElectronApplication) => {
    electronApp.on('console', (data) => console.log(data))
}

/**
 * Get the operating system that the browser is running in.
 * @param page Electron Page instance
 * @returns A promise that will resolve to the Browser's operating system type, or null if
 *     it could not be determined
 */
export const getBrowserOs = async (page: Page) => {
    const userAgent = await page.evaluate(
        async () => window.navigator.userAgent,
    )

    if (userAgent.includes('Windows')) {
        return 'win'
    } else if (userAgent.includes('Mac')) {
        return 'mac'
    } else if (userAgent.includes('X11')) {
        return 'unix'
    } else if (userAgent.includes('Linux')) {
        return 'linux'
    }

    return null
}

/**
 * Launch the Electron application with either the development or production build.
 * @param arg Parameter object
 * @param arg.buildType Determines if the development or the production build is used for testing
 * @param arg.colorScheme Emulates 'prefers-colors-scheme' media feature, supported values are 'light', 'dark', 'no-preference'. See page.emulateMedia([options]) for more details. Passing null resets emulation to system defaults. Defaults to 'light'.
 * @returns A promise that will resolve to an Electron application representation
 */
export const launchElectron = ({
    buildType,
    colorScheme,
}: {
    buildType?: 'development' | 'production'
    colorScheme?: null | 'light' | 'dark' | 'no-preference'
} = {}) => {
    const arg =
        buildType === 'production' ? './build/production/main/main.js' : '.'

    return electron.launch({
        args: [arg],
        colorScheme,
    })
}
