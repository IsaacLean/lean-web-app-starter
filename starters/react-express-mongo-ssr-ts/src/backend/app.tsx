import compression from 'compression'
import express from 'express'
import path from 'path'
import { StrictMode } from 'react'
import { renderToString } from 'react-dom/server'

import {
    PATH_BACKEND_SRC,
    PATH_BUILD_DEV,
    PATH_FRONTEND_BUILD_DEV,
} from '../../PATHS'
import HelloWorld from '../frontend/HelloWorld'

const BUNDLED_BACK_BUILD_PATH = path.resolve(__dirname)
const BUNDLED_FRONT_BUILD_PATH = path.resolve(__dirname, '../frontend')
const BUNDLED_GENERATED_VIEWS_BUILD_PATH = path.resolve(
    __dirname,
    '../generated-views',
)

const app = express()

// Setup EJS templates
app.set('view engine', 'ejs')

// Use the generated views from the frontend build
const viewDirs = []
if (process.env.NODE_ENV === 'test') {
    viewDirs.push(
        `${PATH_BUILD_DEV}/generated-views`,
        `${PATH_BACKEND_SRC}/views`,
    )
} else {
    viewDirs.push(
        BUNDLED_GENERATED_VIEWS_BUILD_PATH,
        `${BUNDLED_BACK_BUILD_PATH}/views`,
    )
}
app.set('views', viewDirs)

// app.use(helmet()) // Middleware that enhances security by setting HTTP response headers

if (process.env.NODE_ENV === 'production') {
    app.use(compression()) // Middleware that compresses most response bodies
}

app.get('/', (req, res, next) => {
    // Server-side render the React app
    let serverSideRendering

    try {
        serverSideRendering = renderToString(
            <StrictMode>
                <HelloWorld />
            </StrictMode>,
        )
    } catch (err) {
        return next(err)
    }

    return res.render('index', {
        title: 'ljas-react-express-mongo-ssr-ts',
        content: serverSideRendering,
    })
})

// Serve the frontend build directory as static files
let frontBuildPath
if (process.env.NODE_ENV === 'test') {
    frontBuildPath = PATH_FRONTEND_BUILD_DEV
} else {
    frontBuildPath = BUNDLED_FRONT_BUILD_PATH
}
app.use('/', express.static(frontBuildPath))

export default app
