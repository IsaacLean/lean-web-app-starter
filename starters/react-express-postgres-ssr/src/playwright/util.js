/**
 * Load the frontend application with either the development or end-to-end host.
 * By default the PLAYWRIGHT_HOST environment variable will be used. If the E2E environment
 * variable is true, then the PLAYWRIGHT_HOST_E2E environment variable will be used instead.
 * You can override these environment variables through some of the available parameters.
 * @param page Playwright Page instance
 * @param arg Parameter object
 * @param arg.gotoOptions Options to pass to Page.goto method
 * @param arg.serverType Determines if the development or end-to-end host is used for testing. Overrides E2E environment variable.
 * @param arg.url Manually set the URL. Overrides all behaviors caused by environment variables.
 * @returns A promise that will resolve to the main resource response
 */
export const loadApp = (page, { gotoOptions, serverType, url } = {}) => {
    let u = process.env.E2E
        ? process.env.PLAYWRIGHT_HOST_E2E
        : process.env.PLAYWRIGHT_HOST
    if (url) {
        u = url
    } else if (process.env.PLAYWRIGHT_HOST_E2E && serverType === 'e2e') {
        u = process.env.PLAYWRIGHT_HOST_E2E
    }

    if (!u) {
        throw new Error('Host was not set')
    }

    return page.goto(u, gotoOptions)
}
