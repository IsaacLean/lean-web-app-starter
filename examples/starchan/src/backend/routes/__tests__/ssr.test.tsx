import {
    fireEvent,
    getQueriesForElement,
    queries,
    render,
    waitForElementToBeRemoved,
} from '@testing-library/react'
import globalJsdom from 'global-jsdom'
import { HttpResponse, http } from 'msw'
import request from 'supertest'

import { buildStore } from '../../../frontend/common/redux'
import { TestApp } from '../../../frontend/common/util/test'
import { server } from '../../../frontend/msw/node'
import app from '../../app'
import { MOCK_THREAD_W_COMMENT } from './MOCK_DATA'
import MOCK_THREAD_LIST_RES from './MOCK_THREAD_LIST_RES.json'

let cleanupJsdom: { (): void }

beforeAll(() =>
    server.listen({
        onUnhandledRequest: (req, print) => {
            const url = new URL(req.url)

            if (url.pathname === '/') {
                // Ignore MSW warning when home page is requested
                return
            }

            print.warning()
        },
    })
)

afterEach(() => {
    server.resetHandlers()
    cleanupJsdom()
})

afterAll(() => server.close())

const MOCK_THREAD_LIST_RESULTS_P1 = MOCK_THREAD_LIST_RES.slice(0, 20)
const MOCK_THREAD_LIST_RESULTS_P2 = MOCK_THREAD_LIST_RES.slice(20, 40)
const MOCK_THREAD_LIST_INFO = {
    hasNextPage: true,
    hasPreviousPage: false,
    totalPages: 3,
}

test('matches snapshot for basic server-side rendering of thread list page', async () => {
    expect.assertions(1)

    const res = await request(app).get('/')

    cleanupJsdom = globalJsdom(res.text, { runScripts: 'dangerously' })

    const rootEl = window.document.getElementById('root')

    if (!rootEl) {
        throw new Error('HTML element with an ID of "root" was not found.')
    }

    const store = buildStore(window.__PRELOADED_STATE__)

    const { asFragment } = render(
        <TestApp initialEntries={['/']} store={store} />,
        { container: rootEl, hydrate: true }
    )

    expect(asFragment()).toMatchSnapshot()
})

test('hydrates empty thread list page', async () => {
    expect.assertions(1)

    const res = await request(app).get('/')

    cleanupJsdom = globalJsdom(res.text, { runScripts: 'dangerously' })

    // When imported directly from Testing Library, screen cannot find
    // document.body from global-jsdom for some reason, so this is a workaround.
    const screen = getQueriesForElement(document.body, queries)

    const rootEl = window.document.getElementById('root')

    if (!rootEl) {
        throw new Error('HTML element with an ID of "root" was not found.')
    }

    const store = buildStore(window.__PRELOADED_STATE__)

    render(<TestApp initialEntries={['/']} store={store} />, {
        container: rootEl,
        hydrate: true,
    })

    expect(
        screen.queryByText(
            /Currently no threads exist. Why don't you create the first one?/i
        )
    ).toBeInTheDocument()
})

test('hydrates thread list page with one page', async () => {
    const MOCK_THREAD_LIST_RESULT = MOCK_THREAD_LIST_RES.slice(0, 5)

    server.use(
        http.get('http://localhost:3000/api/v1/threads', () =>
            HttpResponse.json({
                data: MOCK_THREAD_LIST_RESULT,
                info: {
                    hasNextPage: false,
                    hasPreviousPage: false,
                    totalPages: 1,
                },
            })
        )
    )

    expect.assertions(2)

    const res = await request(app).get('/')

    cleanupJsdom = globalJsdom(res.text, { runScripts: 'dangerously' })

    // When imported directly from Testing Library, screen cannot find
    // document.body from global-jsdom for some reason, so this is a workaround.
    const screen = getQueriesForElement(document.body, queries)

    const rootEl = window.document.getElementById('root')

    if (!rootEl) {
        throw new Error('HTML element with an ID of "root" was not found.')
    }

    const store = buildStore(window.__PRELOADED_STATE__)

    render(<TestApp initialEntries={['/']} store={store} />, {
        container: rootEl,
        hydrate: true,
    })

    // Confirm that the current thread page only shows 5 threads
    expect(screen.getAllByRole('list')[0].children).toHaveLength(5)
    // Confirm that the page navigation only shows 1 page
    expect(screen.getByTestId('page-select').children).toHaveLength(1)
})

test('hydrates thread list page with multiple pages', async () => {
    server.use(
        http.get('http://localhost:3000/api/v1/threads', () =>
            HttpResponse.json({
                data: MOCK_THREAD_LIST_RESULTS_P1,
                info: MOCK_THREAD_LIST_INFO,
            })
        )
    )

    expect.assertions(2)

    const res = await request(app).get('/')

    cleanupJsdom = globalJsdom(res.text, { runScripts: 'dangerously' })

    // When imported directly from Testing Library, screen cannot find
    // document.body from global-jsdom for some reason, so this is a workaround.
    const screen = getQueriesForElement(document.body, queries)

    const rootEl = window.document.getElementById('root')

    if (!rootEl) {
        throw new Error('HTML element with an ID of "root" was not found.')
    }

    const store = buildStore(window.__PRELOADED_STATE__)

    render(<TestApp initialEntries={['/']} store={store} />, {
        container: rootEl,
        hydrate: true,
    })

    // Confirm that the current thread page shows the maximum of 20 threads
    expect(screen.getAllByRole('list')[0].children).toHaveLength(20)
    // Confirm that the page navigation shows 3 pages
    expect(screen.getByTestId('page-select').children).toHaveLength(3)
})

test('thread list page navigates to different page when page select is used', async () => {
    server.use(
        http.get('http://localhost:3000/api/v1/threads', ({ request }) => {
            // Construct a URL instance out of the intercepted request.
            const url = new URL(request.url)
            const page = url.searchParams.get('page')

            if (page === '1') {
                return HttpResponse.json({
                    data: MOCK_THREAD_LIST_RESULTS_P1,
                    info: MOCK_THREAD_LIST_INFO,
                })
            } else if (page === '2') {
                return HttpResponse.json({
                    data: MOCK_THREAD_LIST_RESULTS_P2,
                    info: MOCK_THREAD_LIST_INFO,
                })
            }

            throw new Error('Unsupported mocked page number was requested')
        })
    )

    expect.assertions(4)

    const res = await request(app).get('/')

    cleanupJsdom = globalJsdom(res.text, { runScripts: 'dangerously' })
    window.scrollTo = jest.fn() // Mock scrollTo since jsdom does not implement it

    // When imported directly from Testing Library, screen cannot find
    // document.body from global-jsdom for some reason, so this is a workaround.
    const screen = getQueriesForElement(document.body, queries)

    const rootEl = window.document.getElementById('root')

    if (!rootEl) {
        throw new Error('HTML element with an ID of "root" was not found.')
    }

    const store = buildStore(window.__PRELOADED_STATE__)

    render(<TestApp initialEntries={['/']} store={store} />, {
        container: rootEl,
        hydrate: true,
    })

    // TODO: remove this if unneeded
    // const mockedScrollTo = window.scrollTo as jest.Mock<void>

    // First thread from page 1 should be visible, but first thread from page 2 shouldn't
    expect(
        screen.getByText(MOCK_THREAD_LIST_RESULTS_P1[0].subject)
    ).toBeInTheDocument()
    expect(
        screen.queryByText(MOCK_THREAD_LIST_RESULTS_P2[0].subject)
    ).not.toBeInTheDocument()

    // user-event currently has an issue running in the SSR environment,
    // so fireEvent is used instead
    await fireEvent.click(screen.getByRole('link', { name: '2' }))
    await waitForElementToBeRemoved(() => screen.getByText(/loading.../i))

    // First thread from page 2 should be visible, but first thread from page 1 shouldn't
    expect(
        screen.getByText(MOCK_THREAD_LIST_RESULTS_P2[0].subject)
    ).toBeInTheDocument()
    expect(
        screen.queryByText(MOCK_THREAD_LIST_RESULTS_P1[0].subject)
    ).not.toBeInTheDocument()
})

test('creates new thread with comment when new thread form only has comment', async () => {
    expect.assertions(2)

    const res = await request(app).get('/')

    cleanupJsdom = globalJsdom(res.text, { runScripts: 'dangerously' })

    // When imported directly from Testing Library, screen cannot find
    // document.body from global-jsdom for some reason, so this is a workaround.
    const screen = getQueriesForElement(document.body, queries)

    const rootEl = window.document.getElementById('root')

    if (!rootEl) {
        throw new Error('HTML element with an ID of "root" was not found.')
    }

    const store = buildStore(window.__PRELOADED_STATE__)

    render(<TestApp initialEntries={['/']} store={store} />, {
        container: rootEl,
        hydrate: true,
    })

    // user-event currently has an issue running in the SSR environment,
    // so fireEvent is used instead
    await fireEvent.click(
        screen.getByRole('button', { name: /start a new thread/i })
    )

    // Wait for the new thread form to appear
    const inputComment = (await screen.findByLabelText(
        'Comment'
    )) as HTMLInputElement

    fireEvent.change(inputComment, {
        target: { value: MOCK_THREAD_W_COMMENT.comment },
    })
    await fireEvent.click(screen.getByRole('button', { name: /post/i }))

    await screen.findByRole('button', { name: /post a reply/i })

    // Confirm that the thread page has been navigated to
    expect(
        screen.getByText(`Id.${MOCK_THREAD_W_COMMENT.id}`)
    ).toBeInTheDocument()
    expect(screen.getByText(MOCK_THREAD_W_COMMENT.comment)).toBeInTheDocument()
})
