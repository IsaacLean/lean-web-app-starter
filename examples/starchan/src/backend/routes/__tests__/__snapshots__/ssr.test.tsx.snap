// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot for basic server-side rendering of thread list page 1`] = `
<DocumentFragment>
  <header>
    <h1
      id="title"
    >
      *chan
    </h1>
  </header>
  <hr />
  <main>
    <section
      class="post-form"
    >
      <span
        class="post-form__txt"
      >
        [
        <button
          class="post-form__btn"
        >
          Start a New Thread
        </button>
        ]
      </span>
      <noscript>
        &lt;form action="/" method="post"&gt;&lt;table class="post-form__table"&gt;&lt;tbody&gt;&lt;tr&gt;&lt;th&gt;&lt;label for="subject"&gt;Subject&lt;/label&gt;&lt;/th&gt;&lt;td&gt;&lt;input name="subject" type="text"/&gt;&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;th&gt;&lt;label for="comment"&gt;Comment&lt;/label&gt;&lt;/th&gt;&lt;td&gt;&lt;textarea name="comment"&gt;&lt;/textarea&gt;&lt;/td&gt;&lt;/tr&gt;&lt;tr&gt;&lt;td colSpan="2" class="post-form__post-cell"&gt;&lt;button type="submit"&gt;Post&lt;/button&gt;&lt;/td&gt;&lt;/tr&gt;&lt;/tbody&gt;&lt;/table&gt;&lt;/form&gt;
      </noscript>
    </section>
    <hr />
    <p>
      Currently no threads exist. Why don't you create the first one?
    </p>
  </main>
  <nav>
    <ol
      class="page-select"
      data-testid="page-select"
    >
      <li
        class="page-select__page page-select__page--active"
      >
        [
        <a
          href="/"
        >
          1
        </a>
        ]
      </li>
    </ol>
  </nav>
  <footer
    class="footer"
  >
    <p>
      *chan (pronounced as starchan) is a
      <!-- -->
       
      <a
        href="https://en.wikipedia.org/wiki/Textboard"
      >
        textboard
      </a>
       
      <!-- -->
      that demonstrates one of
      <!-- -->
       
      <a
        href="#"
      >
        Lean JS App Starter's full-stack project setups
      </a>
      . It showcases an Express server that acts as a REST API and a React server-side renderer. These work in tandem with a React frontend.
    </p>
    <nav>
      <ul
        class="footer__links"
      >
        <li>
          <a
            href="#"
          >
            About
          </a>
        </li>
        <li>
          <a
            href="#"
          >
            *chan GitHub
          </a>
        </li>
        <li>
          <a
            href="https://github.com/mattlean/lean-js-app-starter"
          >
            Lean JavaScript Application Starter
          </a>
        </li>
      </ul>
    </nav>
  </footer>
</DocumentFragment>
`;
